GH-2749: Deprecate ChannelInterceptorAware

artembilan opened the pull request:
Fixes https://github.com/spring-projects/spring-integration/issues/2749

The `org.springframework.messaging.support.InterceptableChannel` provides
exact functionality as `ChannelInterceptorAware`

* Make `ChannelInterceptorAware extends InterceptableChannel`
* Suppress deprecation warning whenever we need to keep backward
compatibility
* Fix all other places to deal with `InterceptableChannel` already

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-2749: Deprecate ChannelInterceptorAware

Fixes https://github.com/spring-projects/spring-integration/issues/2749

The `org.springframework.messaging.support.InterceptableChannel` provides
exact functionality as `ChannelInterceptorAware`

* Make `ChannelInterceptorAware extends InterceptableChannel`
* Suppress deprecation warning whenever we need to keep backward
compatibility
* Fix all other places to deal with `InterceptableChannel` already

artembilan commented:
That's just cool to have that `updateCopyrights` task. It saved me a lot of time! 😄

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractMessageChannel.java:
```diff
@@ -243,7 +244,7 @@ public void setMessageConverter(MessageConverter messageConverter) {
 	 */
 	@Override
-	public List<ChannelInterceptor> getChannelInterceptors() {
+	public List<ChannelInterceptor> getInterceptors() {
```
Can't do this; breaking change. Need to deprecate the old method.

artembilan replied:
Unfortunately it clashes with the `InterceptableChannel` API. So, we in chicken-or-egg situation anyway, independently of the version we are going to fix it. Good deal that this method is just `protected`, so visible only for implementors, but not regular end-user code

garyrussell replied:
I've moved it to 5.1.4; let's talk when we have more time.

garyrussell commented on src/reference/asciidoc/channel.adoc:
```diff
@@ -304,6 +304,7 @@ Also, starting with version 5.1, `ChannelInterceptor.postReceive()` is no longer
 Previously, the method was called.
 If you have an interceptor that relies on the previous behavior, implement `afterReceiveCompleted()` instead, since that method is invoked, regardless of whether a message is received or not.
 
+NOTE: Starting with version 5.1.3, the `ChannelInterceptorAware` is deprecated in favor of `InterceptableChannel` from Spring Messaging module, which it extends now for backward compatibility.
```
from **the** Spring Messaging module ... which it **now extends**

garyrussell replied:
`s/5.1.3/5.2/`

from **the** Spring Messaging module

garyrussell reviewed (changes requested):
2 issues; one major, one minor.

artembilan committed:
GH-2749: Deprecate ChannelInterceptorAware

Fixes https://github.com/spring-projects/spring-integration/issues/2749

The `org.springframework.messaging.support.InterceptableChannel` provides
exact functionality as `ChannelInterceptorAware`

* Make `ChannelInterceptorAware extends InterceptableChannel`
* Suppress deprecation warning whenever we need to keep backward
compatibility
* Fix all other places to deal with `InterceptableChannel` already

artembilan commented:
OK. I think its time to forget such a deprecation for `5.1.x` and just apply the change for the current `5.2`.

I'll add a `Migration Guide` note, when the PR is merged.

Thanks

garyrussell reviewed (changes requested):
Minor nit picks

artembilan committed:
* Fix `channel.adoc` for the current version

artembilan commented:
Added Migration Guide note: https://github.com/spring-projects/spring-integration/wiki/Spring-Integration-5.1-to-5.2-Migration-Guide#deprecated-channelinterceptoraware
